### ISSUE TITLE
select(Set, Function) isn't what you might expect

### ISSUE THREAD

**Issue by denham0** [Macaulay2/M2#578]
Check this out:

i837 : select(set{0,1,2},odd)
o837 = set {0, 1, 2}
o837 : Set

i840 : select(set{0,1,2},even)
o840 = set {}
o840 : Set

Now that's not a bug, because a Set is a HashTable, so select applies the function to the values in the hash table.  Which, for a set, are all 1.  But I think most people would expect select(Set, Function) to select a subset (i.e., apply the function to the keys of the hash table, rather than the values.)  It's not a big deal, but I think it would be better if the user didn't have to know how Set was implemented in order to avoid this little trap.

**Comment by DanGrayson** [comment]
Good point!  I think I was thinking of a list as being like a hash table whose keys are 0,1,...,n, as in this example:

```
i1 : f = method();

i3 : f List := v -> new HashTable from apply ( #v, i -> i => v#i );

i5 : f {4,6,11,13}

o5 = HashTable{0 => 4 }
               1 => 6
               2 => 11
               3 => 13

o5 : HashTable

i6 : select (  {4,6,11,13}, even )

o6 = {4, 6}

o6 : List

i7 : select ( f {4,6,11,13}, even )

o7 = HashTable{0 => 4}
               1 => 6

o7 : HashTable
```
We could decree that `select` no longer works for hash tables and introduce new functions `selectKeys` and `selectValues`.  We already do the analogous thing for `apply` and `scan`.

**Comment by denham0** [comment]
That sounds like a good solution.  Then one could use `selectKeys` in order to select from sets as one would from lists.  But `select(Set,Function)` would fail, which should help keep the user out of trouble.  

And it would be nice to select from hash tables both by keys or by values.

**Comment by d-torrance** [comment]
Fixed in #3305
